src/WebJobs.Script.WebHost/Controllers/HostController.cs

```diff
@@ -111,28 +111,33 @@ public async Task<IActionResult> GetHostStatus([FromServices] IScriptHostManager
         [Authorize(Policy = PolicyNames.AdminAuthLevelOrInternal)]
         public async Task<IActionResult> Drain([FromServices] IScriptHostManager scriptHostManager)
         {
-            _logger.LogDebug("Received request for draining host");
-
-            if (!Utility.TryGetHostService(scriptHostManager, out IDrainModeManager drainModeManager))
+            try
             {
-                return StatusCode(StatusCodes.Status503ServiceUnavailable);
-            }
+                _logger.LogDebug("Received request to drain the host");
 
-            await _drainModeSemaphore.WaitAsync();
+                if (!Utility.TryGetHostService(scriptHostManager, out IDrainModeManager drainModeManager))
+                {
+                    return StatusCode(StatusCodes.Status503ServiceUnavailable);
+                }
 
-            // Stop call to some listeners gets stuck, not waiting for the stop call to complete
-            _ = drainModeManager.EnableDrainModeAsync(CancellationToken.None)
-                                .ContinueWith(
-                                    antecedent =>
-                                    {
-                                        if (antecedent.Status == TaskStatus.Faulted)
-                                        {
-                                            _logger.LogError(antecedent.Exception, "Something went wrong enabling drain mode");
-                                        }
+                await _drainModeSemaphore.WaitAsync();
 
-                                        _drainModeSemaphore.Release();
```

Comment:
Thanks for the feedback, this makes a lot of sense. I remember Mathew & Fabio asking me to include a semaphore in the /resume case because we're checking state so I believe that needs to stay. I don't know why we need a semaphore in /drain - maybe @soninaren has context.

Sounds like the fix is to make the semaphores distinct as you suggested, or remove the drain one?

Reply:
Suggest making them distinct.  It doesn't hurt to have a semaphore in drain so we can leave it in there and ask Naren for guidance about removing it.